source/common/router/route_config_update_receiver_impl.cc

```diff
@@ -25,26 +25,50 @@ bool RouteConfigUpdateReceiverImpl::onRdsUpdate(
   vhds_configuration_changed_ = new_vhds_config_hash != last_vhds_config_hash_;
   last_vhds_config_hash_ = new_vhds_config_hash;
   initializeRdsVhosts(route_config_proto_);
-  onUpdateCommon(route_config_proto_, version_info);
-  return true;
-}
 
-void RouteConfigUpdateReceiverImpl::onUpdateCommon(
-    const envoy::config::route::v3::RouteConfiguration& rc, const std::string& version_info) {
   last_config_version_ = version_info;
   last_updated_ = time_source_.systemTime();
   rebuildRouteConfig(rds_virtual_hosts_, vhds_virtual_hosts_, route_config_proto_);
-  config_info_.emplace(RouteConfigProvider::ConfigInfo{rc, last_config_version_});
+  config_info_.emplace(RouteConfigProvider::ConfigInfo{route_config_proto_, last_config_version_});
+
+  config_ = std::make_shared<ConfigImpl>(
+      route_config_proto_, factory_context_,
+      factory_context_.messageValidationContext().dynamicValidationVisitor(), false);
+  return true;
 }
 
 bool RouteConfigUpdateReceiverImpl::onVhdsUpdate(
     const VirtualHostRefVector& added_vhosts, const std::set<std::string>& added_resource_ids,
     const Protobuf::RepeatedPtrField<std::string>& removed_resources,
     const std::string& version_info) {
-  resource_ids_in_last_update_ = added_resource_ids;
+
+  std::map<std::string, envoy::config::route::v3::VirtualHost> vhosts_before_this_update(
+      vhds_virtual_hosts_);
   const bool removed = removeVhosts(vhds_virtual_hosts_, removed_resources);
   const bool updated = updateVhosts(vhds_virtual_hosts_, added_vhosts);
-  onUpdateCommon(route_config_proto_, version_info);
+
+  envoy::config::route::v3::RouteConfiguration route_config_before_this_update;
+  route_config_before_this_update.CopyFrom(route_config_proto_);
+  rebuildRouteConfig(rds_virtual_hosts_, vhds_virtual_hosts_, route_config_proto_);
+  ConfigConstSharedPtr new_config;
+
+  try {
+    new_config = std::make_shared<ConfigImpl>(
+        route_config_proto_, factory_context_,
+        factory_context_.messageValidationContext().dynamicValidationVisitor(), false);
+  } catch (const Envoy::EnvoyException& e) {
+    // revert the changes that failed validation
+    vhds_virtual_hosts_ = vhosts_before_this_update;
+    route_config_proto_.CopyFrom(route_config_before_this_update);
+    throw e;
+  }
+
+  config_ = new_config;
+  last_config_version_ = version_info;
+  last_updated_ = time_source_.systemTime();
```

Comment:
Could we have a common update method with just the minimal amount of common stuff? It's less than before but still some refactor might make sense.

Reply:
done.